[RNMobile] Simplify media insertion flow Part 1 - redux changes

jd-alexander opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->
`gutenberg-mobile` PR https://github.com/wordpress-mobile/gutenberg-mobile/pull/2700

- [Part 1](https://github.com/WordPress/gutenberg/pull/29546) ( Current) 
- [Part 2](https://github.com/WordPress/gutenberg/pull/29547)
- [Part 3](https://github.com/WordPress/gutenberg/pull/29548)

## Description
These changes are related to the PR https://github.com/WordPress/gutenberg/pull/25031 that is being broken down into separate PRs for easier testing and review. 

Once this PR has been approved, we can merge the other two PRs into this one and then merge all the changes into `trunk`. 

The `gutenberg-mobile` PR https://github.com/wordpress-mobile/gutenberg-mobile/pull/2700 contains a temporary ref to the last PR for testing purposes.

The main goal of this PR is to introduce changes to the native redux store behavior that will facilitate the storing of the last block that was inserted i.e the last block the user clicked on the `Inserter` menu. This solution was utilized so that a `Image, Video or Gallery` would be able to query the store for the `clientId` of the last block inserted if there is a match then there can be a resulting action. 


## How has this been tested?

1. These changes were primarily tested using the unit tests that are included within the PR by running `npm run native test`.
2. Subsequent PRs that implement the behavior at a `Component` level will showcase that the functionality works as expected. 

## Types of changes
Changes were made to the native variants of the actions, reducer, and selectors.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

jd-alexander committed:
created actions for adding and clearing last inserted block event.

jd-alexander committed:
added reducer for determining new state based on the action

jd-alexander committed:
added selector to query the state for the last block inserted

guarani commented on packages/editor/src/store/reducer.native.js:
```diff
@@ -80,6 +80,25 @@ export function notices( state = [], action ) {
+ *
+ * @return {Object} Updated state.
+ */
+export function lastBlockInserted( state = {}, action ) {
+	switch ( action.type ) {
+		case 'ADD_LAST_BLOCK_INSERTED':
+			return { ...state, clientId: action.clientId };
+
+		case 'CLEAR_LAST_BLOCK_INSERTED':
+			return {};
+	}
+	return state;
+}
```
Not a blocker at all but I'm curious if state could have been just a plain `clientId` (which is a string or an int IIRC):
```
export function lastBlockInserted( state = null, action ) {
	switch ( action.type ) {
		case 'ADD_LAST_BLOCK_INSERTED':
			return action.clientId;

		case 'CLEAR_LAST_BLOCK_INSERTED':
			return null;
	}
	return state;
}
```
I guess that would have the disadvantage of `lastBlockInserted` not being auto-documenting (like it doesn't say it's a `clientId` anymore). Probably keeping it an object is also more scalable since it's easy to add more attributes without having to refactor.

chipsnyder reviewed (approved):
This looks good to me. I don't have a strong performance, either way, regarding @guarani's [suggestion](https://github.com/WordPress/gutenberg/pull/29546/files#r587635404) so I'll leave that for others to comment on.

guarani commented:
> This looks good to me. I don't have a strong performance, either way, regarding @guarani's [suggestion](https://github.com/WordPress/gutenberg/pull/29546/files#r587635404) so I'll leave that for others to comment on.

I'm totally happy with the way it is, I was just curious 👍 (so don't that comment hold this PR back @jd-alexander! 😄) I'm reviewing the other PRs now.

guarani reviewed (approved):
LGTM :)

guarani commented:
BTW, the failing iOS E2E tests might have been fixed by https://github.com/WordPress/gutenberg/pull/29589, so if you update with the latest from `trunk`, the build might pass 🤞

guarani commented:
> > BTW, the failing iOS E2E tests might have been fixed by #29589, so if you update with the latest from `trunk`, the build might pass 🤞
> 
> There might still be issues with these tests 😞

This [PR](https://github.com/WordPress/gutenberg/pull/29715) looks like it might have fixed the flaky tests, so this might need updating from `trunk`.

jd-alexander committed:
Merge branch 'trunk' into rnmobile/simplify_image_insertion_flow-redux-changes

jd-alexander committed:
[RNMobile] Simplify media insertion flow Part 2 - media upload (#29547)

* added autoOpenMediaUpload prop to the MediaPlaceholder

* Added the auto-opening capabilities to the MediaUpload component.

* Added documentation for the new autoOpenMediaUpload prop

* renamed autoOpenMediaUpload to autoOpen in the MediaUpload component.

* [RNMobile] Simplify media insertion flow - Part 3 component integration (#29548)

* Track the clientId of the block that is inserted.

* implemented auto opening utilizing last block inserted from the store

* added dismissal support for the auto opening picker to the UI tests.

* Updated Dismiss button in closePicker function to Cancel

jd-alexander committed:
Added release notes for auto-opening.

jd-alexander committed:
Merge branch 'trunk' into rnmobile/simplify_image_insertion_flow-redux-changes

jd-alexander commented:
I just noticed the log contains a test failure due to changes made to the web gallery component. The error is in [gallery.test.js](https://github.com/WordPress/gutenberg/blob/rnmobile/simplify_image_insertion_flow-redux-changes/packages/e2e-tests/specs/editor/blocks/gallery.test.js#L13)

```
Expected mock function not to be called but it was called with:
["TypeError: (0 , o(...).wasBlockJustInserted) is not a function  
```

I have started looking into it but I am not too sure what the resolution is because I am not familiar with the web side of e2e testing. I will pull in someone else if I don't see any clues tomorrow.

jd-alexander commented:
Hi @ockham 👋🏾 would you be able to take a look at these [failing e2e test ](https://github.com/WordPress/gutenberg/pull/29546/checks?check_run_id=2127080773)and offer any guidance on how it could be resolved? 

When I was searching for clues in the repo, I saw where you had resolved a [similar issue](https://github.com/WordPress/gutenberg/pull/26574#issuecomment-718927370) a while back. 

The current issue in these set of tests is 


```
Expected mock function not to be called but it was called with:
["TypeError: (0 , o(...).wasBlockJustInserted) is not a function  
```
`wasBlockJustInserted` is a new selector that was added to the store in `.native.js` files. 

Back story: 

This PR introduces changes that allow several media related blocks to auto-open the `MediaUpload` component when inserted. ( Ignore the large diff, this is a base PR that contains two others) One of these blocks is the `gallery` block, and the behavior to make this work was added to the `edit.js` file since, we share it with `native`. Functionality was added to the redux store and used in the gallery block as [seen here.](https://github.com/WordPress/gutenberg/pull/26574#issuecomment-718927370). Since the changes touched the web variant of the gallery component, it seems to have broken several e2e tests that utilize the gallery component. 

I am wondering if there's some form of e2e fixture configurations that contains the jest mocks that can be utilized in these tests but so far I wasn't able to find them.  Let me know if you have any ideas. Thanks much!

youknowriad commented on packages/block-library/src/gallery/edit.js:
```diff
@@ -466,6 +470,22 @@ export default compose( [
+		return {
+			wasBlockJustInserted() {
+				const { clearLastBlockInserted } = dispatch( 'core/editor' );
+				const { wasBlockJustInserted } = select( 'core/editor' );
```
The block-library should ideally not be relying on the  "editor" package because that package is WP specific and the blocks should ideally be generic (this rule is not always 100% respected).


In this situation, it seems, these selectors  are better placed in the block-editor package.

youknowriad replied:
It also doesn't seem like these selectors exist at all in web.

jd-alexander replied:
Understood! Thanks!

ockham commented:
> Hi @ockham 👋🏾 would you be able to take a look at these [failing e2e test ](https://github.com/WordPress/gutenberg/pull/29546/checks?check_run_id=2127080773)and offer any guidance on how it could be resolved?
> 
> When I was searching for clues in the repo, I saw where you had resolved a [similar issue](https://github.com/WordPress/gutenberg/pull/26574#issuecomment-718927370) a while back.
> 
> The current issue in these set of tests is
> 
> ```
> Expected mock function not to be called but it was called with:
> ["TypeError: (0 , o(...).wasBlockJustInserted) is not a function  
> ```
> 
> `wasBlockJustInserted` is a new selector that was added to the store in `.native.js` files.

I'm afraid that the [issue I was facing the other day](https://github.com/WordPress/gutenberg/pull/26574#issuecomment-718927370) is of a fairly different kind. While they both share the "Expected mock function not to be called but it was called with..." part, that only tells us that a function was run that tests didn't expect to be run. In other words, the only thing both issues have in common is the way they are caught/reported. 

In "my" case, that unexpected function call was a block transformation ("Block successfully updated for `%s`"), whereas here, it's an actual `TypeError` in the code. So I'm afraid what helped me track the issue down back then doesn't really apply here :confused:

jd-alexander commented:
Thanks for the feedback @ockham 

> In "my" case, that unexpected function call was a block transformation ("Block successfully updated for %s"), whereas here, it's an actual TypeError in the code. 

This helps, as it confirms the comments made by @youknowriad that the `gallery` block on the web did not expect the function because it is not properly declared there. 

I will work on making the changes needed to get things compliant.

jd-alexander committed:
Merge branch 'trunk' into rnmobile/simplify_image_insertion_flow-redux-changes

jd-alexander commented:
A refactor of the redux changes introduced in this PR is underway here https://github.com/WordPress/gutenberg/pull/30123 Once the unit tests and code cleanups have been completed in that PR it will be merged back into this one, since this base PR targets `trunk`

jd-alexander committed:
[RNMobile] Refactor simplify media flow redux store changes (#30123)

* Moved the last block inserted actions from editor to the block-editor

* Moved the last block inserted reducer from editor to the block-editor

* Moved the last block inserted selector from editor to the block-editor

* Fixed es-lint error.

* Moved last block inserted actions test from editor to the block-editor

* Moved last block inserted reducer test from editor to the block-editor

* Moved last block inserted selector test from editor to the block-editor

* Moved all calls to last block inserted from editor to block-editor

* last block inserter usage in menu native migrated : editor to block-editor

* [RNMobile] Refactor: Simplify media flow redux migration (#30238)

* Add meta argument to insertBlock action

* Add inserter menu source

* Update last block inserted reducer

Instead of using specific actions for tracking the last block inserted, it uses the actions related to the insertion flow.

* Add get last block inserted selector

* Refactor gallery edit component

withSelect and withDispatch logic has been refactored to use useSelect and useDispatch hooks

* Refactor image edit component

wasBlockJustInserted is now calculated with getLastBlockInserted

* Refactor video edit component

wasBlockJustInserted is now calculated with getLastBlockInserted

* Fix reset blocks action in last block inserted reducer

* Add source param to wasBlockJustInserted selector

* Simplify withSelect part of video block

* Removed add/clear last block inserted actions and tests due to refactor

* Removed addLastBlockInserted from the insert menu's onPress.

* rewrote the tests for the lastBlockInserted reducer.

* rewrote tests for wasBlockJustInserted selector.

* optimized clientId

* removed unneeded clientId.

* put the expectedSource inside the test meta object.

* used expectedState insted {} for state related tests.

* used expectedState instead {} for state related tests.

* removed parentheses from describe name.

* return the same state instead of empty state.

* made changes to test name so its intent is clearer.

* made the insertion source optional.

Co-authored-by: Carlos Garcia <fluiddot@gmail.com>

jd-alexander committed:
Merge branch 'trunk' into rnmobile/simplify_image_insertion_flow-redux-changes

jd-alexander committed:
added wasBlockJustInserted prop needed after merge with trunk.

jd-alexander committed:
Merge branch 'trunk' into rnmobile/simplify_image_insertion_flow-redux-changes

jd-alexander commented:
@fluiddot @guarani this is ready for the final review so that it can be merged. @guarani @fluiddot already reviewed the redux changes, and I also reviewed the improvements he made. Nonetheless, you can still give a second set of eyes and share any findings you may have. Apart from that, I think this can be merged once no other issues have been found.

fluiddot commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -1727,6 +1727,31 @@ export function highlightedBlock( state, action ) {
+export function lastBlockInserted( state = {}, action ) {
+	switch ( action.type ) {
+		case 'INSERT_BLOCKS':
+			if ( ! action.updateSelection || ! action.blocks.length ) {
```
After a second thought, I think that we probably shouldn't include the `! action.updateSelection` condition here. The last block inserted should be updated independently on if the block gets selected upon insertion, right?

Originally I thought that this was necessary because we only wanted to show the media picker when the block is selected, but I realised that we're already checking that in the components ([example](https://github.com/WordPress/gutenberg/pull/29546/files#diff-7be1df22c250a7378a0796b6c68da7fc1178a98d9e88ab9f59868a6d0b8c467fR405)), what do you think?

jd-alexander replied:
> After a second thought, I think that we probably shouldn't include the ! action.updateSelection condition here. The last block inserted should be updated independently on if the block gets selected upon insertion, right?

when you say "updated independently on if the block gets selected upon insertion" what exactly do you mean? I want to ensure that I understand you clearly :) 

> Originally I thought that this was necessary because we only wanted to show the media picker when the block is selected, but I realised that we're already checking that in the components (example), what do you think?

Makes sense! I checked the linked example and it went to the `withSelect` in the Gallery component, so I am assuming that you are referring to the `isSelected` usage there. If that's the case, yes, the `isSelected` behavior  was added before the `redux` approach was introduced, so if it's better to keep the selection logic within the reducer, I am not against that, especially if it will be consistent with other behavior in the reducer. 

Another question, I wanted to add to the discussion. In which cases would the block be inserted and it wouldn't be a `selection`. I removed `isSelected` and `updateSelection` from the code in my local environs and the functionality still worked. I will look more into this tomorrow, but I just wanted to share my thoughts here, so you can add any ideas you may have.

fluiddot replied:
>when you say "updated independently on if the block gets selected upon insertion" what exactly do you mean? I want to ensure that I understand you clearly :)

Oh sorry if I wasn't too clear, what I meant is that if this slice of the state contains the last block inserted, it makes sense that it gets updated for any block that is inserted, including the ones with the param `updateSelection` as false. From a consistence POV, I would expect that if I insert a block even if it doesn't get automatically selected (via the `updateSelection` action param), it would be set as the last block inserted.

>Makes sense! I checked the linked example and it went to the withSelect in the Gallery component, so I am assuming that you are referring to the isSelected usage there. If that's the case, yes, the isSelected behavior was added before the redux approach was introduced, so if it's better to keep the selection logic within the reducer, I am not against that, especially if it will be consistent with other behavior in the reducer.

Yeah exactly, I was referring to the `isSelected` usage when calculating the value to pass to `autoOpenMediaUpload` prop.

In my opinion I think it would make more sense to keep the `isSelected` usage as we have currently in the components and update the `lastBlockInserted` reducer.

>Another question, I wanted to add to the discussion. In which cases would the block be inserted and it wouldn't be a selection. I removed isSelected and updateSelection from the code in my local environs and the functionality still worked. I will look more into this tomorrow, but I just wanted to share my thoughts here, so you can add any ideas you may have.

I think it's a uncommon case but looks like there's at least one place where it's being used in the Jetpack project:
https://github.com/Automattic/jetpack/blob/5b637a82a9836213f066809cab550ce3590921cb/projects/plugins/jetpack/extensions/blocks/anchor-fm/editor.js#L48

jd-alexander replied:
> Oh sorry if I wasn't too clear, what I meant is that if this slice of the state contains the last block inserted, it makes sense that it gets updated for any block that is inserted, including the ones with the param updateSelection as false. From a consistence POV, I would expect that if I insert a block even if it doesn't get automatically selected (via the updateSelection action param), it would be set as the last block inserted.

Ah I get what you are saying now. Thanks for the clarification 🙇🏾 

> In my opinion I think it would make more sense to keep the isSelected usage as we have currently in the components and update the lastBlockInserted reducer.

I agree with this as well after gaining a deeper understanding due to your clarifications above. I will make the relevant changes to support this. 

> I think it's a uncommon case but looks like there's at least one place where it's being used in the Jetpack project:
https://github.com/Automattic/jetpack/blob/5b637a82a9836213f066809cab550ce3590921cb/projects/plugins/jetpack/extensions/blocks/anchor-fm/editor.js#L48

Good find 😉 Thanks for checking Jetpack. If I had a sleuthing badge 🔍  I would definitely give you one 😄

fluiddot replied:
>I agree with this as well after gaining a deeper understanding due to your clarifications above. I will make the relevant changes to support this.

Nice! Thanks for making the changes, I'll be watching them so we can have this PR ready as soon as possible.

>Good find 😉 Thanks for checking Jetpack. If I had a sleuthing badge 🔍 I would definitely give you one 😄

Thank you very much! 🕵️😄

fluiddot reviewed (commented):
I tested the changes on Android (_device: Samsung Galaxy S20 FE_) and iOS (_device: iPhone 11_) and works nice!

I didn't approve the PR because I found out a potential change, let me know what you think about it. Apart from that this PR is ready 🟢 , awesome work @jd-alexander 💯 !

jd-alexander committed:
Merge branch 'trunk' into rnmobile/simplify_image_insertion_flow-redux-changes

jd-alexander committed:
removed updateSelection from reducer so it's updated at all times.

jd-alexander commented:
I made the requested changes here bb98f4c @fluiddot so this is ready for another review. 🙇🏾

fluiddot commented on packages/react-native-editor/CHANGELOG.md:
```diff
@@ -10,8 +10,17 @@ For each user feature we should also add a importance categorization label  to i
+-   [*] Image block: Improve text entry for long alt text. [#29670]
+
+## 1.50.0
+
```
Looks like this part is duplicated as per line `25` so I think we could remove these lines.

fluiddot reviewed (approved):
LGTM 🎊 !

_Tested on iPhone 11 (iOS 14.2) and Samsung Galaxy S20 FE 5G (Android 10)._

jd-alexander committed:
Merge branch 'trunk' into rnmobile/simplify_image_insertion_flow-redux-changes
